# SMU Master of Science (Economics) Programming Workshop in Python


## Introduction
In today's class, we will be learning more about Data Cleaning and Manipulation with pandas and NumPy.

---

pandas
![](images/pandas_logo.png "pandas")
pandas is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.

---

NumPy
![](images/numpy_logo.png "NumPy")
NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays.

You can find more information about pandas [here](https://pandas.pydata.org/), and more information about NumPy [here](https://www.numpy.org/).


Before we go dive deep into data cleaning and manipulation, it is useful to take a step back and think about the differences between a function and a method:

1. A function is a "global" function that can be used on different kinds of data structures. For example, the `len` function can be used on lists and strings.
2. A method is a "local" function that can only be used on a specific kind of data structure. For example, the method `pivot_table` can only be used on the DataFrame data structure (we will see this in more detail).

#### Case Study Approach

To learn how we can use pandas and NumPy, we will adopt a case study approach. We have attached 2 datasets (which you can find in the /data folder):

1. The first dataset is from [Kaggle](https://www.kaggle.com/), a machine learning platform. More information about the data can be found [here](https://www.kaggle.com/c/titanic/data). As we will be using this dataset for 3 lessons (including this one), please familiarise yourself with the dataset (the metadata). 

2. The second dataset is a dataset on movie ratings that comes from [GroupLens](https://grouplens.org/). In this dataset, there are 1,000,000 entries on movie ratings by users. The metadata can be found [here](http://files.grouplens.org/datasets/movielens/ml-1m-README.txt). 

---

We will begin our data analysis on the Titanic dataset first. We begin by importing key libraries, such as pandas and NumPy library into Jupyter, by calling the "import" function. In this case, we will only be dealing with the training data.

---

#### Data Manipulation with Titanic
In the first part of data manipulation, we learn how to get summary statistics of the dataset, and the types of data we are dealing with.

if data is the filename,

1. data.head() returns the first five rows of the dataset
2. data.describe() returns the summary statistics of the dataset

In [156]:
import pandas as pd
import numpy as np

# Since data is in the form of a csv file, we can use pd.read_csv
train_data = pd.read_csv('/Users/ngbolin/Desktop/Python/data/titanic/train.csv', index_col = 0) # The first column is the index 
train_data.describe(include='all') # Describe gives a summary of the dataframe 

Survived      Pclass                   Name   Sex         Age  \
count   891.000000  891.000000                    891   891  714.000000   
unique         NaN         NaN                    891     2         NaN   
top            NaN         NaN  O'Connor, Mr. Maurice  male         NaN   
freq           NaN         NaN                      1   577         NaN   
mean      0.383838    2.308642                    NaN   NaN   29.699118   
std       0.486592    0.836071                    NaN   NaN   14.526497   
min       0.000000    1.000000                    NaN   NaN    0.420000   
25%       0.000000    2.000000                    NaN   NaN   20.125000   
50%       0.000000    3.000000                    NaN   NaN   28.000000   
75%       1.000000    3.000000                    NaN   NaN   38.000000   
max       1.000000    3.000000                    NaN   NaN   80.000000   

             SibSp       Parch    Ticket        Fare        Cabin Embarked  
count   891.000000  891.000000       891  891.000000          204      889  
unique         NaN         NaN       681         NaN          147        3  
top            NaN         NaN  CA. 2343         NaN  C23 C25 C27        S  
freq           NaN         NaN         7         NaN            4      644  
mean      0.523008    0.381594       NaN   32.204208          NaN      NaN  
std       1.102743    0.806057       NaN   49.693429          NaN      NaN  
min       0.000000    0.000000       NaN    0.000000          NaN      NaN  
25%       0.000000    0.000000       NaN    7.910400          NaN      NaN  
50%       0.000000    0.000000       NaN   14.454200          NaN      NaN  
75%       1.000000    0.000000       NaN   31.000000          NaN      NaN  
max       8.000000    6.000000       NaN  512.329200          NaN      NaN

The describe( ) method gives us a quick summary of what we are dealing with. For example, we know that there are a total of 12 variables (including passenger ID), and there are a mixture of variables which are categorical ('Ticket', 'Cabin', 'Embarked') as well as numerical ('Pclass', 'Age', 'Fare').

Let's take a look at the first 5 rows of the dataset. We can subset the data using square brackets, [ ], or use the DataFrame method, `head()`.

In [157]:
train_data[0:5]

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

In [158]:
train_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

Now that we have a better idea of what the data looks like, we can conduct some data analysis. Suppose that we are interested in finding out who embarked at location, 'S'. One way to do so is by writing the following code, and it returns all passengers who embarked at the location, 'S'.

In [159]:
train_data[train_data['Embarked'] == 'S'][:10]

Survived  Pclass  \
PassengerId                     
1                   0       3   
3                   1       3   
4                   1       1   
5                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
11                  1       3   
12                  1       1   
13                  0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
7                0      0             17463  51.8625   E46        S  
8                3      1            349909  21.0750   NaN        S  
9                0      2            347742  11.1333   NaN        S  
11               1      1           PP 9549  16.7000    G6        S  
12               0      0            113783  26.5500  C103        S  
13               0      0         A/5. 2151   8.0500   NaN        S

Other than subsetting data on only 1 column, we can also subset data on 2 or more columns.

In [160]:
train_data[(train_data['Embarked'] == 'S') & (train_data['Pclass'] == 1)][:10]

Survived  Pclass                                          Name  \
PassengerId                                                                   
4                   1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)   
7                   0       1                       McCarthy, Mr. Timothy J   
12                  1       1                      Bonnell, Miss. Elizabeth   
24                  1       1                  Sloper, Mr. William Thompson   
28                  0       1                Fortune, Mr. Charles Alexander   
36                  0       1                Holverson, Mr. Alexander Oskar   
56                  1       1                             Woolner, Mr. Hugh   
63                  0       1                   Harris, Mr. Henry Birkhardt   
84                  0       1                       Carrau, Mr. Francisco M   
89                  1       1                    Fortune, Miss. Mabel Helen   

                Sex   Age  SibSp  Parch  Ticket      Fare        Cabin  \
PassengerId                                                              
4            female  35.0      1      0  113803   53.1000         C123   
7              male  54.0      0      0   17463   51.8625          E46   
12           female  58.0      0      0  113783   26.5500         C103   
24             male  28.0      0      0  113788   35.5000           A6   
28             male  19.0      3      2   19950  263.0000  C23 C25 C27   
36             male  42.0      1      0  113789   52.0000          NaN   
56             male   NaN      0      0   19947   35.5000          C52   
63             male  45.0      1      0   36973   83.4750          C83   
84             male  28.0      0      0  113059   47.1000          NaN   
89           female  23.0      3      2   19950  263.0000  C23 C25 C27   

            Embarked  
PassengerId           
4                  S  
7                  S  
12                 S  
24                 S  
28                 S  
36                 S  
56                 S  
63                 S  
84                 S  
89                 S

In [161]:
train_data[train_data['Sex'] == 'male']['Age'].mean()

30.72664459161148

If we want to find the variance of a specific column, we can use the function `np.var`. Similarly, we can find the average fare using the function, `np.mean(train_data['Fare'])`. There exists other functions too, such as `np.median`.

In [162]:
np.var(train_data[(train_data['Embarked'] == 'S') & (train_data['Pclass'] == 1)]['Fare'])

3431.532010935895

In [163]:
print('Average fare is', np.mean(train_data['Fare']))
print('Fare variance is', np.var(train_data['Fare']))

Average fare is 32.2042079685746
Fare variance is 2466.6653116850434


In-class assignment:

1. Previously, we note that the mean survival rate of the dataset is 38.4% (where did we note this?). Suppose you think that passengers who embarked at 'S' has a higher rate of surviving. How do you check this?
2. Are males or females more likely to perish?
3. Which passenger class costs the most?

In [164]:
np.mean(train_data[train_data['Embarked'] == 'S'].Survived)

0.33695652173913043

In [165]:
np.mean(train_data[train_data['Sex'] == 'male'].Survived)

0.18890814558058924

In [166]:
np.mean(train_data[train_data['Sex'] == 'female'].Survived)

0.7420382165605095

In [167]:
for _ in [1, 2, 3]:
    print('Pclass: %d' % _)
    print('Mean Fare: %f' % np.mean(train_data[train_data['Pclass'] == _].Fare))
    print('Median Fare: %f' % np.median(train_data[train_data['Pclass'] == _].Fare))
    print()

Pclass: 1
Mean Fare: 84.154687
Median Fare: 60.287500

Pclass: 2
Mean Fare: 20.662183
Median Fare: 14.250000

Pclass: 3
Mean Fare: 13.675550
Median Fare: 8.050000



#### More Data Manipulation with Titanic!
Now that we have a basic understanding of data manipulation in Python, we can proceed with slightly more difficult notations:

1. Locating and imputing missing values in the data
2. Pivot Tables
3. Preliminary Text Cleaning
4. Groupby

---

#### 1. Locating and Imputing Missing Values
The DataFrame method, `apply`, takes a function as its argument, and applys the function to the whole dataframe. In the code cell below, we apply the function `pd.isnull()` to the dataset. Note that if the cell contains True, it has a missing value.

In [168]:
np.sum(pd.isnull(train_data))

Survived      0
Pclass        0
Name          0
Sex           0
Age         177
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

In [169]:
train_data.shape

(891, 11)

Data Cleaning:

It appears that many of the Age and Cabin data are missing. In this case, we can do the following, depending on the column category:

1. Fill the mean value (with `df.fill_na()`)
2. Create a new category
3. Drop the columns

---

The DataFrame method `fillna()`, fills all missing values with an integer, float, or string provided by the user. Setting the argument, `inplace=True` results in the dataframe being altered **forever**.

In [170]:
# Filling in mean value for Age
def fill_mean_value(train_data, column): 
    train_data[column].fillna(train_data[column].mean(), inplace=True)
    return train_data

new_train_data = fill_mean_value(train_data, 'Age')

In [171]:
# Creating a new category
def add_new_category(train_data, column, new_cat_name):
    train_data[column].fillna(new_cat_name, inplace=True)
    return train_data

new_train_data = add_new_category(new_train_data, 'Cabin', 'No Cabin')
new_train_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare     Cabin Embarked  
PassengerId                                                              
1                1      0         A/5 21171   7.2500  No Cabin        S  
2                1      0          PC 17599  71.2833       C85        C  
3                0      0  STON/O2. 3101282   7.9250  No Cabin        S  
4                1      0            113803  53.1000      C123        S  
5                0      0            373450   8.0500  No Cabin        S

def drop_columns(train_data, column):
    columns = [col for col in train_data.columns if col != column]
    return train_data[columns]

new_train_data = drop_columns(new_train_data, 'Cabin')
new_train_data.head()

del new_train_data['Cabin']

To include a portion on imputing conditional age based on Sex, Parch and SibSp

#### 2. Pivot Tables

We can use the method, `pivot_table` to generate conditional summary statistics. For example, one might be interested in the number of male or female passengers who survived, and embarked at different locations. In that case, one can use the following code:

In [172]:
new_train_data.pivot_table(values='Survived', index='Sex', columns=['Embarked'], aggfunc='mean')

Embarked         C         Q         S
Sex                                   
female    0.876712  0.750000  0.689655
male      0.305263  0.073171  0.174603

In-class assignment:

1. Using the DataFrame method, `pivot_table`, find the average fare paid by female passengers who survived.
2. Using the DataFrame method, `pivot_table`, find the average age of male passengers who were of passenger class 1.

In [173]:
## Your code here
new_train_data.pivot_table(values='Fare', index='Survived', columns=['Sex'], aggfunc='mean')

Sex          female       male
Survived                      
0         23.024385  21.960993
1         51.938573  40.821484

In [174]:
## Your code here
new_train_data.pivot_table(values='Fare', index='Sex', columns=['Pclass'], aggfunc='mean')

Pclass           1          2          3
Sex                                     
female  106.125798  21.970121  16.118810
male     67.226127  19.741782  12.661633

In [175]:
## We can add in new conditional columns
new_train_data.pivot_table(values='Fare', index='Survived', columns=['Sex', 'Pclass'], aggfunc='mean')

Sex           female                            male                      
Pclass             1          2          3         1          2          3
Survived                                                                  
0         110.604167  18.250000  19.773093  62.89491  19.488965  12.204469
1         105.978159  22.288989  12.464526  74.63732  21.095100  15.579696

Let's take a closer look at the "Ticket" column of the dataset.

In [176]:
new_train_data['Ticket'].head()

PassengerId
1           A/5 21171
2            PC 17599
3    STON/O2. 3101282
4              113803
5              373450
Name: Ticket, dtype: object

In [177]:
new_train_data['Ticket'].head()

PassengerId
1           A/5 21171
2            PC 17599
3    STON/O2. 3101282
4              113803
5              373450
Name: Ticket, dtype: object

In [178]:
new_train = new_train_data['Ticket'].loc[1]
new_train.split(' ')

['A/5', '21171']

##### Functional/Lambda Programming

Note that the ticket column contains both integers and characters. We can define a simple function that separates on the space. This is a good time to introduce lambda programming. A lambda function is a small anonymous function, which can take any number of arguments, but can only have one expression. 


We won't go through lambda nor functional programming in detail, since this is best reserved for a slightly more advanced class.

In [179]:
new_train_data['Ticket'].apply(lambda x: x.split(' '))

PassengerId
1             [A/5, 21171]
2              [PC, 17599]
3      [STON/O2., 3101282]
4                 [113803]
5                 [373450]
6                 [330877]
7                  [17463]
8                 [349909]
9                 [347742]
10                [237736]
11              [PP, 9549]
12                [113783]
13            [A/5., 2151]
14                [347082]
15                [350406]
16                [248706]
17                [382652]
18                [244373]
19                [345763]
20                  [2649]
21                [239865]
22                [248698]
23                [330923]
24                [113788]
25                [349909]
26                [347077]
27                  [2631]
28                 [19950]
29                [330959]
30                [349216]
              ...         
862                [28134]
863                [17466]
864            [CA., 2343]
865               [233866]
866               [236852]
867       [SC/PA

Suppose we are only interested in the integer component of the ticket code. One way to get it is through the following:

In [180]:
x = ['a', 'b', 'c']
x[-3]

'a'

In [181]:
new_train_data['Ticket'].apply(lambda x: x.split(' ')[-1]).head()

PassengerId
1      21171
2      17599
3    3101282
4     113803
5     373450
Name: Ticket, dtype: object

#### 3. Preliminary Text Analysis
What if, instead, we are interested in the string component of the ticket code? How can we get it? One way of deriving the string component of the code is through the following:

In [182]:
def get_string_component_from_ticket(data, column='Ticket'):
    string_lst = []
    for row in new_train_data['Ticket'].apply(lambda x: x.split(' ')):
        if len(row) > 1: string_lst.append(row[0])
        else: string_lst.append('Null')
    return string_lst
        
get_string_component_from_ticket(new_train_data)[:5]

['A/5', 'PC', 'STON/O2.', 'Null', 'Null']

Using the previous 2 chunks of codes, we can dissect the Ticket column into 2 components: the string component and the integer component. Let's write some code to do that:

In [183]:
def dissect_ticket_column(data, column='Ticket'):
    int_component = data[column].apply(lambda x: x.split(' ')[-1])
    str_component = get_string_component_from_ticket(new_train_data)
    
    data['Ticket_Int'] = int_component
    data['Ticket_Str'] = str_component
    
    return data

new_data = dissect_ticket_column(new_train_data)
new_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare     Cabin Embarked  \
PassengerId                                                               
1                1      0         A/5 21171   7.2500  No Cabin        S   
2                1      0          PC 17599  71.2833       C85        C   
3                0      0  STON/O2. 3101282   7.9250  No Cabin        S   
4                1      0            113803  53.1000      C123        S   
5                0      0            373450   8.0500  No Cabin        S   

            Ticket_Int Ticket_Str  
PassengerId                        
1                21171        A/5  
2                17599         PC  
3              3101282   STON/O2.  
4               113803       Null  
5               373450       Null

Looks like everything worked out fine! We have successfully split the ticket column into 2 different columns for us to carry out analysis on. It might be useful to conduct a similar analysis on the column, Name.

---

In-class assignment (You can define your own functions, or use functional programming):
1. Split the name column of new_data in to 2 different columns, Last_Name, and First_Name. Hint: You can split on the comma punctuation (', ').  
2. After splitting, do another split on the First_Name column on the following string: ". " (note the space) to obtain the honorific name (Mr, Ms etc.) and the actual Last_Name. Name the columns: Honorific and True_Last_Name.

In [184]:
def dissect_name_column(data, column='Name'):
    ## Your code here
    data['Last_Name'] = data[column].apply(lambda x: x.split(", ")[0])
    data['First_Name'] = data[column].apply(lambda x: x.split(", ")[1])
    
    data['Honorific'] = data['First_Name'].apply(lambda x: x.split(". ")[0])
    data['True_Last_Name'] = data['First_Name'].apply(lambda x: x.split(". ")[1])
    return data

new_data = dissect_name_column(new_data)
new_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare     Cabin Embarked  \
PassengerId                                                               
1                1      0         A/5 21171   7.2500  No Cabin        S   
2                1      0          PC 17599  71.2833       C85        C   
3                0      0  STON/O2. 3101282   7.9250  No Cabin        S   
4                1      0            113803  53.1000      C123        S   
5                0      0            373450   8.0500  No Cabin        S   

            Ticket_Int Ticket_Str  Last_Name  \
PassengerId                                    
1                21171        A/5     Braund   
2                17599         PC    Cumings   
3              3101282   STON/O2.  Heikkinen   
4               113803       Null   Futrelle   
5               373450       Null      Allen   

                                             First_Name Honorific  \
PassengerId                                                         
1                                       Mr. Owen Harris        Mr   
2            Mrs. John Bradley (Florence Briggs Thayer)       Mrs   
3                                           Miss. Laina      Miss   
4                    Mrs. Jacques Heath (Lily May Peel)       Mrs   
5                                     Mr. William Henry        Mr   

                                    True_Last_Name  
PassengerId                                         
1                                      Owen Harris  
2            John Bradley (Florence Briggs Thayer)  
3                                            Laina  
4                    Jacques Heath (Lily May Peel)  
5                                    William Henry

In [185]:
np.unique(new_data['Honorific'])

array(['Capt', 'Col', 'Don', 'Dr', 'Jonkheer', 'Lady', 'Major', 'Master',
       'Miss', 'Mlle', 'Mme', 'Mr', 'Mrs', 'Ms', 'Rev', 'Sir',
       'the Countess'], dtype=object)

Going forward, it is useful to check what kinds of tickets (Ticket_Str) there are. To do so, we can use the `np.unique` function to do that (what other functions could we have used?). Let's do that.

In [186]:
np.unique(new_data['Ticket_Str'])

array(['A./5.', 'A.5.', 'A/4', 'A/4.', 'A/5', 'A/5.', 'A/S', 'A4.', 'C',
       'C.A.', 'C.A./SOTON', 'CA', 'CA.', 'F.C.', 'F.C.C.', 'Fa', 'Null',
       'P/PP', 'PC', 'PP', 'S.C./A.4.', 'S.C./PARIS', 'S.O./P.P.',
       'S.O.C.', 'S.O.P.', 'S.P.', 'S.W./PP', 'SC', 'SC/AH', 'SC/PARIS',
       'SC/Paris', 'SCO/W', 'SO/C', 'SOTON/O.Q.', 'SOTON/O2', 'SOTON/OQ',
       'STON/O', 'STON/O2.', 'SW/PP', 'W./C.', 'W.E.P.', 'W/C', 'WE/P'],
      dtype=object)

It appears that the data is slightly dirty - there appears to be many mistakes (we actually do not know if these are genuine mistakes or not). One way to verify this is to count the number of times each category appears. If a category only appears once, and is lexically close to another category, then it is very likely to be a mistake.

One way to do so is through the use of a dictionary, looping over every value and keeping count of them. Alternatively, we can use the Series method, `value_counts`. A Series is basically a column of the DataFrame.

In [187]:
new_data['Ticket_Str'].value_counts()

Null          665
PC             60
C.A.           27
STON/O         12
A/5            10
W./C.           9
CA.             8
SOTON/O.Q.      8
SOTON/OQ        7
A/5.            7
CA              6
STON/O2.        6
F.C.C.          5
S.O.C.          5
C               5
SC/PARIS        5
SC/Paris        4
S.O./P.P.       3
A/4             3
A/4.            3
PP              3
SC/AH           3
S.C./PARIS      2
P/PP            2
SOTON/O2        2
A./5.           2
WE/P            2
A.5.            2
SCO/W           1
A/S             1
SW/PP           1
Fa              1
S.O.P.          1
W.E.P.          1
W/C             1
SO/C            1
F.C.            1
A4.             1
SC              1
S.W./PP         1
S.C./A.4.       1
S.P.            1
C.A./SOTON      1
Name: Ticket_Str, dtype: int64

In-class Assignment:

Use the dictionary method to generate the same table above. Helper code has been provided below.

In [188]:
def unique_values_dict(data, column):
    ticket_dict = {}
    for ticket_value in new_data[column].values:
        if ticket_value not in ticket_dict:
            ticket_dict[ticket_value] = 1
        else:
            ticket_dict[ticket_value] += 1
    return ticket_dict

unique_values_dict(new_data, "Ticket_Str")

{'A./5.': 2,
 'A.5.': 2,
 'A/4': 3,
 'A/4.': 3,
 'A/5': 10,
 'A/5.': 7,
 'A/S': 1,
 'A4.': 1,
 'C': 5,
 'C.A.': 27,
 'C.A./SOTON': 1,
 'CA': 6,
 'CA.': 8,
 'F.C.': 1,
 'F.C.C.': 5,
 'Fa': 1,
 'Null': 665,
 'P/PP': 2,
 'PC': 60,
 'PP': 3,
 'S.C./A.4.': 1,
 'S.C./PARIS': 2,
 'S.O./P.P.': 3,
 'S.O.C.': 5,
 'S.O.P.': 1,
 'S.P.': 1,
 'S.W./PP': 1,
 'SC': 1,
 'SC/AH': 3,
 'SC/PARIS': 5,
 'SC/Paris': 4,
 'SCO/W': 1,
 'SO/C': 1,
 'SOTON/O.Q.': 8,
 'SOTON/O2': 2,
 'SOTON/OQ': 7,
 'STON/O': 12,
 'STON/O2.': 6,
 'SW/PP': 1,
 'W./C.': 9,
 'W.E.P.': 1,
 'W/C': 1,
 'WE/P': 2}

There appears to be quite a number of mistakes, based on our analysis. For example, A./5., A.5., A/5 and A/5. seem to be referring to the same cabin type. Same goes for the types A/4 and A/4 and A/4.. To clean the data, we can write a simple function that removes all kinds of punctuations in the data.

Let's search for ["clean punctuations python"](https://www.google.com/search?q=clean+punctuations+python&oq=clean+punctuations+python&aqs=chrome..69i57j0l4.2954j0j1&sourceid=chrome&ie=UTF-8) using Google's search engine.

Google recommends using the library string, and module punctuation. Let's find out what's in `string.punctuation first'.

In [189]:
import string

string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

The following code was taken from [StackOverflow](https://stackoverflow.com/questions/265960/best-way-to-strip-punctuation-from-a-string). Essentially, the code takes some data structure as its input, loops over each element in the data structure, and removes all punctuation from the string.

In [190]:
def clean_punctuations(data, column):
    clean_data = []
    for row in data[column].values:
        clean_data.append(row.translate(str.maketrans('', '', string.punctuation)))
    return clean_data

print(clean_punctuations(new_data, 'Ticket_Str')[:10])

['A5', 'PC', 'STONO2', 'Null', 'Null', 'Null', 'Null', 'Null', 'Null', 'Null']


All punctuation marks have been removed from the `Ticket_Str` column! This is a neat trick that you can use on any specific datasets to remove puncutations or any other character that are uninformative. Now that we have cleaned up the dataset, let's take a look at the remaining unique values in the data.

In [191]:
print(len(np.unique(new_data['Ticket_Str'])))
print(len(np.unique(clean_puncutations(new_data, 'Ticket_Str'))))

43
30


In [192]:
# Convert to lower case
new_data['Ticket_Str'].apply(lambda x: x.lower()).head()

PassengerId
1         a/5
2          pc
3    ston/o2.
4        null
5        null
Name: Ticket_Str, dtype: object

Let's take a look at the missing values in the data once again. Note that we have 2 observations with missing values.

In [193]:
np.sum(pd.isnull(new_data))

Survived          0
Pclass            0
Name              0
Sex               0
Age               0
SibSp             0
Parch             0
Ticket            0
Fare              0
Cabin             0
Embarked          2
Ticket_Int        0
Ticket_Str        0
Last_Name         0
First_Name        0
Honorific         0
True_Last_Name    0
dtype: int64

When we are faced with such a scenario, one thing we might do is to to input the most common value in the dataset. To do this, we can use the `describe` method on the dataframe once again.

In [194]:
new_data.describe(include='all')

Survived      Pclass                   Name   Sex         Age  \
count   891.000000  891.000000                    891   891  891.000000   
unique         NaN         NaN                    891     2         NaN   
top            NaN         NaN  O'Connor, Mr. Maurice  male         NaN   
freq           NaN         NaN                      1   577         NaN   
mean      0.383838    2.308642                    NaN   NaN   29.699118   
std       0.486592    0.836071                    NaN   NaN   13.002015   
min       0.000000    1.000000                    NaN   NaN    0.420000   
25%       0.000000    2.000000                    NaN   NaN   22.000000   
50%       0.000000    3.000000                    NaN   NaN   29.699118   
75%       1.000000    3.000000                    NaN   NaN   35.000000   
max       1.000000    3.000000                    NaN   NaN   80.000000   

             SibSp       Parch    Ticket        Fare     Cabin Embarked  \
count   891.000000  891.000000       891  891.000000       891      889   
unique         NaN         NaN       681         NaN       148        3   
top            NaN         NaN  CA. 2343         NaN  No Cabin        S   
freq           NaN         NaN         7         NaN       687      644   
mean      0.523008    0.381594       NaN   32.204208       NaN      NaN   
std       1.102743    0.806057       NaN   49.693429       NaN      NaN   
min       0.000000    0.000000       NaN    0.000000       NaN      NaN   
25%       0.000000    0.000000       NaN    7.910400       NaN      NaN   
50%       0.000000    0.000000       NaN   14.454200       NaN      NaN   
75%       1.000000    0.000000       NaN   31.000000       NaN      NaN   
max       8.000000    6.000000       NaN  512.329200       NaN      NaN   

       Ticket_Int Ticket_Str  Last_Name First_Name Honorific True_Last_Name  
count         891        891        891        891       891            891  
unique        679         43        667        803        17            799  
top          1601       Null  Andersson   Mr. John        Mr           John  
freq            7        665          9          7       517              9  
mean          NaN        NaN        NaN        NaN       NaN            NaN  
std           NaN        NaN        NaN        NaN       NaN            NaN  
min           NaN        NaN        NaN        NaN       NaN            NaN  
25%           NaN        NaN        NaN        NaN       NaN            NaN  
50%           NaN        NaN        NaN        NaN       NaN            NaN  
75%           NaN        NaN        NaN        NaN       NaN            NaN  
max           NaN        NaN        NaN        NaN       NaN            NaN

In [195]:
new_data.describe(include='all')['Embarked'].top

'S'

For the Embarked column, there are 3 unique values, with "S" as the most common value, appearing 644 out of 889 times.

---

In-class assignment:

1. Define a function that takes data and column as its input, and returns the most common value. You can assume that the column provided is a categorical or string column.
2. Define another function that fills the NA value with the most common value (refer to first function), and returns the data.

In [196]:
## Your code here
def most_common_value(data, column):
    return new_data.describe(include='all')[column].top

def fill_na_with_mcv(data, column):
    data[column].fillna(most_common_value(data, column), inplace=True)
    return data

In [197]:
new_data = fill_na_with_mcv(new_data, 'Embarked')

In [199]:
np.sum(pd.isnull(new_data))

Survived          0
Pclass            0
Name              0
Sex               0
Age               0
SibSp             0
Parch             0
Ticket            0
Fare              0
Cabin             0
Embarked          0
Ticket_Int        0
Ticket_Str        0
Last_Name         0
First_Name        0
Honorific         0
True_Last_Name    0
dtype: int64

Finally, we can conduct some meaningful analysis. One question we might ask is what factors are highly correlated with the Survival factor? To answer this question, we can check the correlation across different features, using the `corr` method on `new_data`. It returns the correlation coefficient between any 2 features in the dataframe. Note that the correlation has to lie between -1 and 1, and the larger the magnitude of the coefficient, the more correlated the 2 features are.

In [32]:
new_data.corr()

Survived    Pclass       Age     SibSp     Parch      Fare
Survived  1.000000 -0.338481 -0.069809 -0.035322  0.081629  0.257307
Pclass   -0.338481  1.000000 -0.331339  0.083081  0.018443 -0.549500
Age      -0.069809 -0.331339  1.000000 -0.232625 -0.179191  0.091566
SibSp    -0.035322  0.083081 -0.232625  1.000000  0.414838  0.159651
Parch     0.081629  0.018443 -0.179191  0.414838  1.000000  0.216225
Fare      0.257307 -0.549500  0.091566  0.159651  0.216225  1.000000

From the correlation matrix, it appears that the features, Age, Fare and Passenger Class are highly correlated with the Survival Probability of an individual. However, we also note that the features, Sex and Embarked do not show up in this matrix. This is due to the fact that they are categorical or ordinal factors. However, we can still use the `pivot_table` method to find out the likelihood of a person surviving given a particular gender.

Below, we give 3 different ways to plot the relationship between survival likelihood against gender and embarkation point.

In [33]:
new_data.pivot_table(values='Survived', index='Sex', aggfunc='mean')

Survived
Sex             
female  0.742038
male    0.188908

In [201]:
new_data.pivot_table(values='Survived', index='Pclass', columns='Sex', aggfunc='mean')

Sex       female      male
Pclass                    
1       0.968085  0.368852
2       0.921053  0.157407
3       0.500000  0.135447

In [35]:
new_data.pivot_table(values='Survived', index='Sex', columns='Embarked', aggfunc='mean')

Embarked         C         Q         S
Sex                                   
female    0.876712  0.750000  0.689655
male      0.305263  0.073171  0.174603

As it turns out, Females have a very high rate of survival. This might be due to many reasons (they might be younger, or they may have paid a higher fare for their tickets etc.). Let's explore this phenomenon in larger detail. We can use the `pivot_table` method to do so.

---

In-class assignment:
1. Test our hypothesis that females are younger than males on the Titanic on average.
2. Test our hypothesis that females pay more than males on average.

In [203]:
## Your code here
new_data.pivot_table(values='Age', index='Sex', aggfunc='mean')
new_data.pivot_table(values='Fare', index='Sex', aggfunc='mean')

Fare
Sex              
female  44.479818
male    25.523893

#### 4. Groupby
One other thing that we can do with DataFrames is the `groupby` method. From the [groupby article on pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html), a groupby operation involves some combination of splitting the object, applying a function, and combining the results. This can be used to group large amounts of data and compute operations on these groups.

---

To motivate an example, let's take a closer look at the "Ticket" column of the DataFrame. There are 2 ways to subset a column: `new_data['Ticket']` and `new_data.Ticket`. We prefer the first method since it allows for "spaces".

In [204]:
len(np.unique(new_data['Ticket']))

681

In [205]:
len(new_data['Ticket'])

891

We note that some of the passengers shared tickets: while there were only 681 unique tickets, there were a total of 891 passengers. This begs the question: who are the passengers who shared these tickets, and how many tickets were shared? 

As it turns out, we can answer this question using the DataFrame method, `new_data.groupby`. In our case, we can create a new Series that returns 1 when the ticket is shared by only 1 passenger, and 2 if it is shared by 2 passengers etc.

In [213]:
# The following function says the following: we want to group the DataFrame by the column, Ticket.
groupby_tickets = new_data.groupby('Ticket')

# Then, we are interested in only the counts of the Name column
shared_tickets = groupby_tickets['Name'].transform('count')
shared_tickets.head()

PassengerId
1    1
2    1
3    1
4    2
5    1
Name: Name, dtype: int64

It turns out that passenger 4 is sharing the ticket with someone else.

---

In-class assignment: Who is passenger 4 sharing the ticket with?

In [221]:
## Your code here
ticket_number = new_data['Ticket'].iloc[3]
new_data[new_data['Ticket'] == ticket_number]

Survived  Pclass                                          Name  \
PassengerId                                                                   
4                   1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)   
138                 0       1                   Futrelle, Mr. Jacques Heath   

                Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  \
PassengerId                                                            
4            female  35.0      1      0  113803  53.1  C123        S   
138            male  37.0      1      0  113803  53.1  C123        S   

            Ticket_Int Ticket_Str Last_Name  \
PassengerId                                   
4               113803       Null  Futrelle   
138             113803       Null  Futrelle   

                                     First_Name Honorific  \
PassengerId                                                 
4            Mrs. Jacques Heath (Lily May Peel)       Mrs   
138                           Mr. Jacques Heath        Mr   

                            True_Last_Name  
PassengerId                                 
4            Jacques Heath (Lily May Peel)  
138                          Jacques Heath

Now that you have a working understanding of the DataFrame method, `groupby`, use it to solve the following questions:

1. Several passengers are sharing a cabin with one another. Use the DataFrame method, `groupby` to create a Series that returns how many passengers are sharing the same cabin.
2. Use the "First Name" column to create a Series that returns the number of passengers that are sharing the same First Name.

In [227]:
# Question 1
## Your code here
shared_cabin1 = new_data.groupby('Cabin')['Name'].transform('count')

## Alternative code - gives the same result
groupby_cabin = new_data.groupby('Cabin')
shared_cabin2 = groupby_cabin['Name'].transform('count')

In [233]:
# Question 2
## Your code here
shared_first_name = new_data.groupby('First_Name')['Name'].transform('count')
shared_first_name

PassengerId
1      1
2      1
3      1
4      1
5      4
6      7
7      1
8      1
9      1
10     1
11     1
12     2
13     4
14     1
15     1
16     1
17     1
18     1
19     1
20     1
21     1
22     2
23     1
24     1
25     1
26     1
27     1
28     1
29     2
30     2
      ..
862    1
863    1
864    1
865    1
866    1
867    1
868    1
869    1
870    1
871    1
872    1
873    1
874    3
875    1
876    1
877    1
878    1
879    1
880    1
881    1
882    1
883    1
884    1
885    1
886    1
887    1
888    1
889    1
890    1
891    3
Name: Name, Length: 891, dtype: int64

In [269]:
new_data.groupby('Ticket')['Fare'].transform('mean')

PassengerId
1        7.25000
2       71.28330
3        7.92500
4       53.10000
5        8.05000
6        8.45830
7       51.86250
8       21.07500
9       11.13330
10      30.07080
11      16.70000
12      26.55000
13       8.05000
14      31.27500
15       7.85420
16      16.00000
17      29.12500
18      13.00000
19      18.00000
20       7.22500
21      26.00000
22      13.00000
23       8.02920
24      35.50000
25      21.07500
26      31.38750
27       7.22500
28     263.00000
29       7.87920
30       7.89580
         ...    
862     11.50000
863     25.92920
864     69.55000
865     13.00000
866     13.00000
867     13.85830
868     50.49580
869      9.50000
870     11.13330
871      7.89580
872     52.55420
873      5.00000
874      9.00000
875     24.00000
876      7.22500
877      9.53125
878      7.89580
879      7.89580
880     83.15830
881     26.00000
882      7.89580
883     10.51670
884     10.50000
885      7.05000
886     29.12500
887     13.00000
888     30.00000
88

Finally, we can save the dataset using the method `.to_csv()`. 

In [234]:
new_data.to_csv('data/newdata.csv')

In [235]:
new_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare     Cabin Embarked  \
PassengerId                                                               
1                1      0         A/5 21171   7.2500  No Cabin        S   
2                1      0          PC 17599  71.2833       C85        C   
3                0      0  STON/O2. 3101282   7.9250  No Cabin        S   
4                1      0            113803  53.1000      C123        S   
5                0      0            373450   8.0500  No Cabin        S   

            Ticket_Int Ticket_Str  Last_Name  \
PassengerId                                    
1                21171        A/5     Braund   
2                17599         PC    Cumings   
3              3101282   STON/O2.  Heikkinen   
4               113803       Null   Futrelle   
5               373450       Null      Allen   

                                             First_Name Honorific  \
PassengerId                                                         
1                                       Mr. Owen Harris        Mr   
2            Mrs. John Bradley (Florence Briggs Thayer)       Mrs   
3                                           Miss. Laina      Miss   
4                    Mrs. Jacques Heath (Lily May Peel)       Mrs   
5                                     Mr. William Henry        Mr   

                                    True_Last_Name  
PassengerId                                         
1                                      Owen Harris  
2            John Bradley (Florence Briggs Thayer)  
3                                            Laina  
4                    Jacques Heath (Lily May Peel)  
5                                    William Henry

### Data Manipulation with MovieLens

The following example is taken from the book, "Python for Data Analysis" written by Wes McKinney, the inventor of Pandas. If you're interested in buying the book, you can find it [here](https://www.bookdepository.com/Python-for-Data-Analysis-2e-Wes-McKinney/9781491957660?ref=grid-view&qid=1564376975506&sr=1-1). I derive no royalties or commissions.

If you've taken a closer look at the MovieLens dataset, you would have noticed that the extension for the dataset is ".dat". However, it turns out that we can still use the `pd.read_csv` function for this. Alternatively, we can use the `pd.read_table` function for this as well. pandas has the ability to read different kinds of data files, and ".dat" is one of them. 

The MovieLens dataset has 3 different tables: 

1. "unames", a table that contains the usernames, gender, age and occupation etc. of users
2. "rnames", a table that contains the ratings, movie ids and timestamp etc.
3. "mnames", a table that contains the movie titles, movie ids and genres of the movies

We can add an argument, `names` to the functions `pd.read_table` or `pd.read_csv` to tell Python which table we are interested in.

In [237]:
# Import user data table
unames = ['user_id', 'gender', 'age', 'occupation', 'zip']
users = pd.read_csv('data/movielens/ml-1m/users.dat', engine='python', sep='::', header=None, names=unames)

# Import ratings table
rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_csv('data/movielens/ml-1m/ratings.dat', engine='python', sep='::', header=None, names=rnames)

# Import mnames table
mnames = ['movie_id', 'title', 'genres']
movies = pd.read_csv('data/movielens/ml-1m/movies.dat', engine='python', sep='::', header=None, names=mnames)

It is almost canon to take a look at the headers of the dataframes you have read after you have imported them into Python, so that's what we'll do in the next few code blocks.

In [242]:
users.head()

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455

In [243]:
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

In [244]:
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

A keen observer would have noted that there are some dataframes that we can combine using columns that we can merge on. For example, one can merge the users and rating dataframes using the columns, user_id. Alternatively, one can also merge the ratings and movies dataset using the column, movie_id. Below, we merge the users and ratings dataframes.

In [247]:
dat = pd.merge(users, ratings, on='user_id')

Let's take a look at the merged dataframe. Following the "general" principle of data analysis, we take a look at the first 5 rows of the dataframe.

In [248]:
dat.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp
0        1      F    1          10  48067      1193       5  978300760
1        1      F    1          10  48067       661       3  978302109
2        1      F    1          10  48067       914       3  978301968
3        1      F    1          10  48067      3408       4  978300275
4        1      F    1          10  48067      2355       5  978824291

We can now merge this dataframe with the movies dataframe using the column, movie_id. Let's do that.

In [249]:
dat = pd.merge(dat, movies, on='movie_id')

In [254]:
dat.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

We managed to combine the 3 dataframes into 1, using the function `pd.merge`. This function takes as argument 2 dataframes to combine together, as well as some other arguments.

In [255]:
?pd.merge

Here are some following questions we can ask:

1. What are the most popular titles?
2. Which user rated the most movies? 
3. Which users are likely to give a very high rating on the movies that they rate?
4. What areas are users most likely to stay at?
5. Do older users tend to watch different genres of movies from the younger users?

It turns out that we can answer these questions fairly easily with the use of Python, Pandas and Numpy. Below, we answer each question sequentially.

In [261]:
# Popular Titles
def top_n_titles(n):
    # .size() returns the counts of the title.
    titles = dat.groupby('title').size()
    return titles.sort_values(ascending=False)[:n]

top_n_titles(10)

title
American Beauty (1999)                                   3428
Star Wars: Episode IV - A New Hope (1977)                2991
Star Wars: Episode V - The Empire Strikes Back (1980)    2990
Star Wars: Episode VI - Return of the Jedi (1983)        2883
Jurassic Park (1993)                                     2672
Saving Private Ryan (1998)                               2653
Terminator 2: Judgment Day (1991)                        2649
Matrix, The (1999)                                       2590
Back to the Future (1985)                                2583
Silence of the Lambs, The (1991)                         2578
dtype: int64

It appears that most of the users are fans of the movies, American Beauty, Star Wars and Jurassic Park. Now, which users rated the most movies? We can define a function that returns these users.

In [54]:
def most_active_users(n):
    active_users = dat.groupby('user_id').size()
    return active_users.sort_values(ascending=False)[:n]

most_active_users(10)

user_id
4169    2314
1680    1850
4277    1743
1941    1595
1181    1521
889     1518
3618    1344
2063    1323
1150    1302
1015    1286
dtype: int64

Of all the questions, the third question is the most important ones. Chances are, when you're doing programming, you're more likely to be interested in these types of questions (conditional questions).

In [264]:
def user_average_ratings(n):
    user_ratings = dat.groupby('user_id')['rating'].mean()
    return user_ratings.sort_values(ascending=False)[:n]

user_average_ratings(10)

user_id
283     4.962963
2339    4.956522
3324    4.904762
3902    4.890909
446     4.843137
447     4.837838
4649    4.818182
4634    4.813725
1131    4.796117
4925    4.761905
Name: rating, dtype: float64

It appears that users 4169 and 1680 rated the most movies out of all the users. In addition, it appears that user 283 and 2339 are very lenient with their ratings, as their average rating comes to about 4.95 and above.

In [56]:
def most_populous_area(n):
    user_zip = dat.groupby('zip').size()
    return user_zip.sort_values(ascending=False)[:n]

most_populous_area(10)

zip
94110    3802
60640    3430
98103    3204
95616    3079
02138    3019
55408    2787
48135    2725
97401    2663
10025    2632
10024    2594
dtype: int64

Checking the zip codes, it appears that most users stay in the area of San Francisco and Illinois. Let's find the average age of the movie reviewers.

In [57]:
dat.age.mean()

29.73831369243828

On average, the age of a reviewer is about 30. Let's categories the older reviewers as 1.5 times the average age, 45.

In [276]:
dat['older'] = dat.age > 45
dat.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  older  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  False  
1  One Flew Over the Cuckoo's Nest (1975)  Drama   True  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  False  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  False  
4  One Flew Over the Cuckoo's Nest (1975)  Drama   True

In [277]:
def most_popular_genres(n):
    old_df = dat[dat.older == True]
    young_df = dat[dat.older == False]
    
    old_genres = old_df.groupby('genres').size()
    young_genres = young_df.groupby('genres').size()
    
    top_old_genres = old_genres.sort_values(ascending=False)[:n]
    top_young_genres = young_genres.sort_values(ascending=False)[:n]
    
    return top_old_genres, top_young_genres

top_old_genres, top_young_genres = most_popular_genres(10)

In [279]:
top_young_genres

genres
Comedy                     106570
Drama                       95632
Comedy|Romance              37950
Comedy|Drama                37501
Drama|Romance               25370
Action|Thriller             24409
Horror                      20788
Drama|Thriller              16116
Action|Adventure|Sci-Fi     16081
Thriller                    15877
dtype: int64

In [61]:
top_young_genres

genres
Comedy                     106570
Drama                       95632
Comedy|Romance              37950
Comedy|Drama                37501
Drama|Romance               25370
Action|Thriller             24409
Horror                      20788
Drama|Thriller              16116
Action|Adventure|Sci-Fi     16081
Thriller                    15877
dtype: int64

Consider the following problem:

We are working with the Academy Awards (Oscars) to select the top films of the year by genre. We have the following criteria for selection:

1. For each category, we are interested in movies with 250 ratings and above.
2. For each selected film, our tolerance for 1-star ratings is 1%.

In the following code block, 4 functions have been defined to help us achieve our objective of selecting the top movie that fulfils both requirements:

1. `popular_movies` - This function takes 2 arguments, a dataframe and an integer, and returns a dataframe of movies with more than n ratings.
2. `good_review_movies` - This function takes 2 arguments, a dataframe and a floating point, and returns movies with number of 1-star ratings < x%.
3. `best_movies` - This function takes 2 arguments, a dataframe and the genre, and returns the top movie for the genre given the dataframe.
4. `oscar_winners` - This function takes 3 arguments, a dataframe, an integer and a floating point, and returns a dictionary with genres of movies as the keys and the movie names as the values.

In [80]:
def popular_movies(data, n=250):
    '''
    Input: 
        1. data, a pd.DataFrame
        2. n, an integer
    
    Output: movies, pd.DataFrame
    
    Returns movies with more than n ratings
    '''
    # Groupby movies data to get movies that haveAT than n ratings
    movies_groupby = data.groupby('title').size()
    greater_n_ratings = np.where(data.groupby('title').size() >= n)
    top_movies = list(movies_groupby.iloc[greater_n_ratings].index)
    
    # After getting these top movies, create a new column that returns True if the movie is a top movie
    data['top_movies'] = data.title.apply(lambda x: x in top_movies)
    return data[data.top_movies == True]

def good_reviews_movies(data, tolerance=0.01):
    '''
    Input: 
        1. data, a pd.DataFrame
        2. tolerance, a floating point
    
    Output: good_movies, a pd.DataFrame
    
    Returns movies with number of 1-star rating < 1%
    '''
    # Group the movie ratings by the number of ratings each movie has
    movie_ratings_count = data.groupby('title')['movie_id'].count()
    
    # Subset dataframe into users who rated 1-star
    poor_df = data[data.rating == 1]
    poor_ratings_count = poor_df.groupby('title')['movie_id'].count()
    
    # Check for the proportion of 1-star ratings across all ratings
    tolerance_list = dict()
    for movie in movie_ratings_count.index:
        if movie in poor_ratings_count.index:
            tolerance_list[movie] = poor_ratings_count.loc[movie]/movie_ratings_count.loc[movie]
        else:
            tolerance_list[movie] = 0
    
    # Check if each movie passes tolerance rate
    good_review_movies = []
    for movie in tolerance_list.keys():
        if tolerance_list[movie] < tolerance: good_review_movies.append(movie)
    
    data['good_movies'] = data.title.apply(lambda x: x in good_review_movies)
    return data

def best_movie(data, genre):
    '''
    Input:
        1. data, a pd.DataFrame
        2. genre, a string
        
    Output:
        1. top_movie, a string
    
    Returns the top movie given a genre.
    '''
    # Subsets data by genre and calculates average ratings
    subset = data[data.genres == genre]
    movie_ratings = subset.groupby('title')['rating'].mean()
    
    # Returns top movie based on ratings
    top_movie = movie_ratings.sort_values(ascending=False).index[0]
    return top_movie
    
def oscar_winners(data, n=250, tolerance=0.01):
    '''
    Input:
        1. data, a pd.DataFrame
        2. n, an integer for the minimum number of ratings
        3. tolerance rate, a floating point, for proportion of 1-star ratings
        
    Output: top_movies, a dicttionary, key-value pairs are genre: movies
    
    Returns the Oscar winners for each category.
    '''
    # Get top movies with at least n reviews and has less than 1% of 1-star ratings
    movies_with_n_ratings = popular_movies(data, n)
    movies_with_good_reviews = good_reviews_movies(movies_with_n_ratings, tolerance)
    
    # Generate the Oscar Winners (based on our conditions)
    top_movies = dict()
    for genre in np.unique(data.genres.values):
        try: 
            oscar = best_movie(movies_with_good_reviews, genre)
            top_movies[genre] = oscar
        except: pass
    return top_movies

After writing our helper functions, let's proceed to find the Oscar Winners for each category!

In [81]:
oscar_winners = oscar_winners(dat)

/Users/ngbolin/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:50: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


Let's take a look at the Oscar winners!

In [82]:
oscar_winners

{'Action': 'Goldfinger (1964)',
 'Action|Adventure': 'Raiders of the Lost Ark (1981)',
 'Action|Adventure|Animation': 'Princess Mononoke, The (Mononoke Hime) (1997)',
 'Action|Adventure|Animation|Horror|Sci-Fi': 'Heavy Metal (1981)',
 "Action|Adventure|Children's|Comedy": 'Inspector Gadget (1999)',
 "Action|Adventure|Children's|Sci-Fi": 'Super Mario Bros. (1993)',
 'Action|Adventure|Comedy': 'Three Musketeers, The (1993)',
 'Action|Adventure|Comedy|Crime': 'Midnight Run (1988)',
 'Action|Adventure|Comedy|Horror': 'Evil Dead II (Dead By Dawn) (1987)',
 'Action|Adventure|Comedy|Horror|Sci-Fi': 'Army of Darkness (1993)',
 'Action|Adventure|Comedy|Romance': 'Princess Bride, The (1987)',
 'Action|Adventure|Comedy|Sci-Fi': 'Men in Black (1997)',
 'Action|Adventure|Crime': 'Rumble in the Bronx (1995)',
 'Action|Adventure|Crime|Drama': 'Batman (1989)',
 'Action|Adventure|Drama': 'Ben-Hur (1959)',
 'Action|Adventure|Drama|Romance': 'First Knight (1995)',
 'Action|Adventure|Drama|Sci-Fi|War': 'S

In-class assignment:

Using the dictionary, `oscar_winners`, answer the following questions:
1. Which movies won the most Oscars?
2. Which movie, out of all the Oscar Winners, had the highest rating?
3. How many movies won more than 1 Oscar?

In [83]:
## Your code here

In [84]:
## Your code here

In [85]:
## Your code here